book/styles.md

```diff
@@ -724,104 +806,137 @@ def style(node, rules):
+``` {.python indent=4}
+def recurse(self, node):
+    if isinstance(node, Text):
+        self.text(node)
+    else:
+        # ...
 ```
 
-Now support for the `i`, `b`, `small`, and `big` tags can all be moved
-to CSS:
+Since the `line` field now stores color, we need to modify `flush`,
+which reads that variable:
 
-``` {.css}
-i { font-style: italic; }
-b { font-weight: bold; }
-small { font-size: 12px; }
-big { font-size: 20px; }
+``` {.python indent=4}
+def flush(self):
+    if not self.line: return
+    metrics = [font.metrics() for x, word, font, color in self.line]
+    # ...
+    for x, word, font, color in self.line:
+        # ...
+        self.display_list.append((x, y, word, font, color))
+    # ...
 ```
 
-Another place where the code depends on specific tag names is
-`has_block_children`, which relies on a list of inline elements. The
-CSS `display`, which can be either `block` or `inline`, replaces that
-mechanism.[^lots-of-values] So we can add all the inline elements to
-our browser style sheet:
+Finally, that `display_list` is read in `draw`:
 
-[^lots-of-values]: Modern CSS adds way more values, like `run-in` or
-    `inline-block` or `flex` or `grid`, and it has layout modes set by
-    other properties, like `float` and `position`. These values allow
-    for layouts you couldn't do with just `block` and `inline`, and
-    people really care about making their web pages look good.
-
-``` {.css}
-a { display: inline; }
-em { display: inline; }
-/* ... */
+``` {.python indent=4}
+def draw(self, display_list):
+    # ...
+    for x, y, word, font, color in self.display_list:
+        display_list.append(DrawText(x, y, word, font, color))
 ```
 
-And then read that in `has_block_children`:[^default-block]
+Phew! That was a lot of coordinated changes, so please test everything
+and make sure it works. You should now see links on this page appear
+in blue---and you'll also notice that the rest of the text has become
+slightly lighter.^[Check out [the book's stylesheet](book.css) to see
+the details.]
 
-[^default-block]: Actually, the `display` property is specified to
-    default to `inline`, not `block`, so a real browser would list all
-    the block elements in its browser stylesheet. But there are a lot
-    more of them, so I'm cutting a corner here.
+Font Properties
+===============
 
-``` {.python}
-def has_block_children(self):
-    for child in self.node.children:
-        # ...
-        elif child.style.get("display", "inline") == "inline":
-            return False
-    return True
+Well---since we're already mucking around with `InlineLayout`, let's
+also modify it to look up the font size, weight, and style information
+via CSS instead of using the `style`, `weigth`, and `size` fields on
+`InlineLayout`:
+
+::: {.todo}
+How do we resolve percentage values for nodes? And the `font-size`
+needs to be converted from pixels to points.[^72ppi]
+:::
+
+[^72ppi]: Normally you think of points as a physical length unit (one
```

Comment:
This differs between Windows and Mac.

Reply:
I thought the difference here would be handled by Tk. We should test.

Reply:
This is a blog post that goes into what I'm talking about:

https://tonsky.me/blog/font-size/


Read after "why the number 72?